File: src/main/java/raylras/zen/model/type/IntType.java
package raylras.zen.model.type;

import raylras.zen.model.symbol.Operator;
import raylras.zen.model.symbol.Symbol;
import raylras.zen.model.symbol.SymbolFactory;

import java.util.List;

public class IntType extends NumberType {

    public static final IntType INSTANCE = new IntType();

    @Override
    public List<Symbol> getSymbols() {
        return SymbolFactory.builtinSymbols()
                .add(super.getSymbols())
                .operator(Operator.RANGE, IntRangeType.INSTANCE, params -> params.parameter("val", this))
                .build();
    }

    @Override
    public String toString() {
        return "int";
    }

}

File: core/src/main/kotlin/raylras/intellizen/type/IntType.kt
package raylras.intellizen.type

import raylras.intellizen.CompilationEnvironment
import raylras.intellizen.symbol.Operator
import raylras.intellizen.symbol.Symbol
import raylras.intellizen.symbol.symbolSequence

object IntType : NumberType {
    override val typeName = "int"

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        return symbolSequence {
            operator(Operator.RANGE, IntRangeType) {
                parameter("value", IntType)
            }
        } + super.getSymbols(env)
    }

    override fun toString() = typeName
}
